WinRT.Runtime/Projections/IEnumerable.net5.cs

```diff
@@ -0,0 +1,638 @@
+                    //  2. This allows us to preserve the same semantics as generic collection iteration when iterating
+                    //     beyond the end of the collection - namely that Current continues to return the last value
+                    //     of the collection
+                    if (m_hadCurrent)
+                    {
+                        m_current = _iterator._Current;
+                    }
+                }
+                catch (Exception e)
+                {
+                    // Translate E_CHANGED_STATE into an InvalidOperationException for an updated enumeration
+                    if (Marshal.GetHRForException(e) == ExceptionHelpers.E_CHANGED_STATE)
+                    {
+                        throw new InvalidOperationException(ErrorStrings.InvalidOperation_EnumFailedVersion);
+                    }
+                    else
+                    {
+                        throw;
+                    }
+                }
+
+                return m_hadCurrent;
+            }
+
+            public void Reset()
+            {
+                throw new NotSupportedException();
+            }
+
+            public void Dispose()
+            {
+            }
+        }
+
+        public sealed class ToAbiHelper : global::Windows.Foundation.Collections.IIterator<T>, IBindableIterator
+        {
+            private readonly global::System.Collections.Generic.IEnumerator<T> m_enumerator;
+            private bool m_firstItem = true;
+            private bool m_hasCurrent;
+
+            internal ToAbiHelper(global::System.Collections.Generic.IEnumerator<T> enumerator) => m_enumerator = enumerator;
+
+            public T _Current
+            {
+                get
+                {
+                    // IEnumerator starts at item -1, while IIterators start at item 0.  Therefore, if this is the
+                    // first access to the iterator we need to advance to the first item.
+                    if (m_firstItem)
+                    {
+                        m_firstItem = false;
+                        _MoveNext();
+                    }
+
+                    if (!m_hasCurrent)
+                    {
+                        ExceptionHelpers.ThrowExceptionForHR(ExceptionHelpers.E_BOUNDS);
+                    }
+
+                    return m_enumerator.Current;
+                }
+            }
+
+            public bool HasCurrent
+            {
+                get
+                {
+                    // IEnumerator starts at item -1, while IIterators start at item 0.  Therefore, if this is the
+                    // first access to the iterator we need to advance to the first item.
+                    if (m_firstItem)
+                    {
+                        m_firstItem = false;
+                        _MoveNext();
+                    }
+
+                    return m_hasCurrent;
+                }
+            }
+
+            public bool _MoveNext()
+            {
```

Comment:
Is it possible for someone to call movenext first without calling current or hascurrent or is it just us calling it?  If it is possible, would that make m_firstItem wrong?